llvm/include/llvm/IR/IntrinsicsAMDGPU.td

```diff
@@ -3210,4 +3210,12 @@ def int_amdgcn_addrspacecast_nonnull : DefaultAttrsIntrinsic<
   [llvm_anyptr_ty], [llvm_anyptr_ty],
   [IntrNoMem, IntrSpeculatable]
 >;
+
+// This intrinsic is used to track the user SGPRs that hold implicit kernel
+// arguments. The i32 argument for this intrinsic represents the offset in
+// bytes from the value pointed to by the implicitarg pointer.
+def int_amdgcn_preload_implicitarg :
```

Comment:
Fundamentally implicit arguments aren't different from a user argument. Why do these require a special intrinsic to track them if we already handle the not-user case? If we do need this intrinsic, I think in terms of patch splitting it would be better to introduce this separately with a dedicated test showing it works for the full range of legal types

Reply:
I'd like to avoid needing an intrinsic. The issue is how do I identify the loads that use the implicit_arg pointer to properly allocate user SGPRs before isel, and then also identify and replace these loads during isel. The best idea I've come up with is to check ahead of time and use an intrinsic.

We allocate user SGPRs, then system SGPRs directly after. If it is possible to delay the allocation of system SGPRs until the end of isel then it's doable. What do you think?

Reply:
This flow sounds backwards. You shouldn't need to identify the loads, they shouldn't appear in CodeGen in the first place. I would expect the IR pass marking the inreg arguments to replace the load uses with the direct argument use. The argument lowering would then just emit the CopyFromReg from the correct position for the inreg cases (instead of the legacy emit load lowering we have now)

Reply:
Am I understanding correctly that you think implicit hidden arguments should be added to the function signature in some way? Or what do you mean by "IR pass marking the inreg arguments to replace the load uses with the direct argument use". Maybe that should just be done in the frontend instead of just an IR pass since that is where this lowering to loads is currently happening. I mean it makes sense to me that they should be actual arguments but I'm not sure about the consequences of doing that.

Reply:
Gentle ping. Could you expand on what you mean here?

Reply:
I forgot this was specifically about the implicit arguments, but essentially yes. You could rewrite the IR signature of the function, giving you uniformity of representation, although it's a bit cumbersome to do that (you have to create a new function with the new type, then splice the body, and steal the name)

I have a long term desire to have clang only emit byref for every kernel argument, and avoid the triple handling in AMDPULowerKernelArguments and the 2 selector implementations. We would then have a uniformity of representation where non-byref arguments are passed in registers. The blocker before is you lose some optimizations, from attributes that traditionally can only be applied to arguments. The main one was noalias, but we have metadata now that works in the load case, so it's probably a reasonable time to revisit that project.

Reply:
I believe rewriting the function signature in this way would be an ABI breaking change that would require coordination beyond the compiler.

Do you think that the current approach in this PR could be feasible in the meantime, or do you think that changing the handling of implicit arguments in this way is something we should move towards right away?